### PR TITLE
Bonima codereview2

### PR DESCRIPTION
**Pull request by bonima21**
_**Analysis of the program.**_

**Was the program available in UC Github on time?**
Yes it was 

**Is the program documented/commented well enough for you to understand?**
Yes it was 
**Does the program compile?**
The program does compile 

**Links to three specific commits that you made to your own group's GitHub repository after Sprint 1's deadline, and before Sprint 2's deadline.**  

https://github.com/jakeFarrish/MedMapper/commit/541dfd71a199352343866adb9f09581dbf9af6ad
https://github.com/jakeFarrish/MedMapper/commit/ca0dbef2b05ab5bca484404a8f7fc0e23c1507c9
https://github.com/jakeFarrish/MedMapper/commit/e9f87f070c38bb17f2c191161cf0cec498d78967

Three specific technical concepts that you learned from this code review.  
How to use avoid using heavy operations.
How to use 'liveData.observeAsState()'
how to add and use column

Citations:
https://kotlinlang.org/docs/reference/coding-conventions.html
https://developer.android.com/distribute/best-practice
https://docs.sonarqube.org/display/SONARQUBE45/Developers%27+Seven+Deadly+Sins

### CODE REVIEW

File: app/src/main/java/com/example/placefinder/MainActivity.kt [hunk 1]
```diff
@@ -36,15 +36,15 @@ class MainActivity : ComponentActivity() {
         override fun onCreate(savedInstanceState: Bundle?) {
             super.onCreate(savedInstanceState)
             setContent {
-                viewModel.fetchParks()
-                val parks by viewModel.parks.observeAsState(initial = emptyList())
+                var parks by remember { mutableStateOf(emptyList<Address>()) }
```
Comment (bonima21) [inline app/src/main/java/com/example/placefinder/MainActivity.kt:39]: In the given code, viewModel.fetchParks() method is being called on the main thread. This method could take a considerable amount of time to complete, which could freeze the UI thread and make the app unresponsive. To avoid this, I moved this operation to a background thread.

File: app/src/main/java/com/example/placefinder/MainActivity.kt [hunk 2]
```diff
@@ -99,13 +99,22 @@ class MainActivity : ComponentActivity() {
...
 
+    @Composable
+    fun ParksList(parks: List<Address>) {
+        LazyColumn {
+            items(parks) { park ->
+                Parks(park)
+            }
+        }
+    }
+
```
Comment (bonima21) [inline app/src/main/java/com/example/placefinder/MainActivity.kt:117]: In the given code, the Parks() composable is using the Row() and Box() composable to display the parks. However, if the number of parks is large, it can lead to performance issues. Instead, you they should use the LazyColumn() composable which loads only the visible items and recycles the others.

File: app/src/main/java/com/example/placefinder/FirstFragment.kt [hunk 3]
```diff
@@ -24,8 +24,8 @@ class FirstFragment : Fragment() {
             inflater: LayoutInflater, container: ViewGroup?,
             savedInstanceState: Bundle?
     ): View? {
+        val binding = FragmentFirstBinding.inflate(inflater, container, false)
 
-        _binding = FragmentFirstBinding.inflate(inflater, container, false)
         return binding.root
```
Comment (bonima21) [inline app/src/main/java/com/example/placefinder/FirstFragment.kt:29]: In the current implementation, the code accesses views through the _binding property which can be null at any time before onDestroyView() is called. This can lead to crashes if the view is accessed when it is already destroyed.

Comment (discospiff) [reply app/src/main/java/com/example/placefinder/FirstFragment.kt:29]: Yeah... it took me a minute to sort this one out.  A few thoughts:

* Force unwrap is an absolute no-no:   private val binding get() = _binding!!
* And there is a possible null here, which could create a NullPointerException: private var _binding: FragmentFirstBinding? = null

On the other hand, the proposed solution has a shadow condition: same variable name used for two different variables, in two different scopes:

private val binding get() = _binding!!

 val binding = FragmentFirstBinding.inflate(inflater, container, false)

Bottom line, I recommend that the group review and refactor this carefully, as there are several items here that are not best practice, introduce technical debt, and are error prone.

File: app/src/main/java/com/example/placefinder/MainActivity.kt (hunk 1 above)
Comment (discospiff) [reply app/src/main/java/com/example/placefinder/MainActivity.kt:39]: LaunchedEffect to run coroutines in compose... good one!

File: app/src/main/java/com/example/placefinder/MainActivity.kt (hunk 2 above)
Comment (discospiff) [reply app/src/main/java/com/example/placefinder/MainActivity.kt:117]: I don't see where this function, ParksList, is getting called, so I am not sure if this function is used... but yes, I agree.  Better to use LazyColumn than a series of Row() and Box() elements.

**Comment by discospiff** [comment]
This is a very deep code review.  I studied it all for a good while.  It has some well considered suggestions.
I recommend that the team review the suggestions, run this branch and ensure it meets their needs, and then merge.
